### PR TITLE
support nested forms isolated from parent

### PR DESCRIPTION
**Pull request by gonzaloruizdevilla**

### CODE REVIEW

**Comment by awerlang** [comment]
Missing a test for submit event's, which would fail because HTML nested structure would eventually fire a submit event on topmost form (structurally speaking), even from a control inside isolated form.

**Comment by petebacondarwin** [comment]
RE naming: root and isolate already have other meanings in angular, and this case doesn't not map accurately to those meanings. I would be happy with `topForm` or better `topLevelForm`.

**Comment by petebacondarwin** [comment]
Also drop the `ng-form-options` since `ng-model-options` is actually a directive that can appear anywhere  and applies to all descendents and this is a single attribute for a single form.

So something like:

```html
<ng-form ng-form-top-level="false">
```

**Comment by gonzaloruizdevilla** [comment]
@awerlang good catch!
@petebacondarwin I will try to change the name and add the event test next weekend.

**Comment by petebacondarwin** [comment]
Let's see if we can get this in 1.5

**Comment by gkalpak** [comment]
This probably needs a rebase since f8a07dd, 290b504 and c6110e8.

**Comment by petebacondarwin** [comment]
Fancy moving this one forward @gkalpak ?

**Comment by gkalpak** [comment]
@petebacondarwin,  sure thing !
<sub>Moving stuff forward is the new fancy :)</sub>

**Comment by petebacondarwin** [comment]
@gonzaloruizdevilla I have landed your first two commits to close off their related issues.
The last commit is not passing on Travis. Can you take a look?

**Comment by gonzaloruizdevilla** [comment]
sure, I'll try to find some time tonight

File: src/ng/directive/form.js [hunk 1]
```diff
@@ -497,18 +507,27 @@ var formDirectiveFactory = function(isNgForm) {
                 event.preventDefault();
               };
 
+              var handleKeypress = function(event) {
```
Comment (Narretz) [inline src/ng/directive/form.js]: That's too simple, see https://docs.angularjs.org/api/ng/directive/form#submitting-a-form-and-preventing-the-default-action
Maybe we should instead listen on the submit event and see if the first parent form of explicitOriginalTarget is 1) not the current submit target and 2) has the do-not-propagate attribute set

**Comment by Narretz** [comment]
I would like to get this in, but I'm not a fan of the ngFormTopLevel attribute. It really doesn't convey what it does very well. I'd prefer something like ng-form-no-parent, or ng-form-no-child, ng-form-detached.

File: src/ng/directive/form.js (hunk 1 above)
Comment (Narretz) [inline src/ng/directive/form.js]: Actually, the whole submit handling seems problematic. In `ngForm`, we don't emulate the submit behavior of regular forms. So in order to be able to "submit" a detached form, we would need to emulate the submit behavior (as requested in https://github.com/angular/angular.js/issues/2513). The current patch doesn't achieve this, and without it, I am not sure how useful the detached form is.

**Comment by Narretz** [comment]
This is not gonna happen in core due to the reasons outlined in https://github.com/angular/angular.js/pull/10193#discussion_r66080985

**Pull request closed without merging**